Cache EnrichCache CacheKey hashCode

joegallo opened the pull request:
Related to #96050

The `org.elasticsearch.common.cache.Cache` calls into `hashCode` quite a bit, sometimes while holding one or more locks. That's fine as long as calculating the `hashCode` is very fast, but in the case of `EnrichCache.CacheKey` it isn't, see #96050.

By pre-computing and storing the hash, we can ensure that subsequent `hashCode` calls are instantaneous, and by doing it in eagerly on object construction we can be certain that the calculation is being done outside any of the `org.elasticsearch.common.cache.Cache` locking.

joegallo committed:
Remove unused method

joegallo committed:
This can be a record

joegallo committed:
Precompute the hashCode for the CacheKey

joegallo committed:
Add some flavor text

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

elasticsearchmachine commented:
Hi @joegallo, I've created a changelog YAML for you.

joegallo committed:
Update docs/changelog/96332.yaml

joegallo commented:
Note to self that we need to use the [`Co-authored-by` syntax](https://github.blog/2018-01-29-commit-together-with-co-authors/) when merging this so that @jbaiera gets the proper attribution.

joegallo commented:
![Screenshot 2023-05-24 at 12 30 19 PM](https://github.com/elastic/elasticsearch/assets/187034/85058431-65e1-475e-a809-6f3a966cc729)

Here's five calls to `hashCode` in the bad version of the flamegraph, this is the case of us needing to put something into the cache and to evict something else afterwards (because we've exceeded the maximum size of the cache due to the put).

joegallo commented:
And here's the flamegraph pair for the cache hit side of things:

![Screenshot 2023-05-24 at 12 46 22 PM](https://github.com/elastic/elasticsearch/assets/187034/ceb70ff5-f434-4284-bcda-3499157d2fd8)

![Screenshot 2023-05-24 at 12 47 09 PM](https://github.com/elastic/elasticsearch/assets/187034/baa2244b-de6a-4e97-87bc-9a119d6f2514)

We replace 2 to 5 `hashCode` calls (top graph, left hand side, purple) in the before with 1 `computeHashCode` call (bottom graph, right hand side, purple) in the after.

Once again, comparing apples to apples, the `EnricheCache.get` time dropped from 6.85% in the before to 3.44% in the after.

masseyke reviewed (approved):
Seems like a change that at worst will not make things any slower, and at best might speed things up a little. As we discussed, we'd all feel more comfortable if we had an enrich performance benchmark.

joegallo committed:
This can be final

Less a performance enhancment, more a statement of intent. It's so
close to being able to be a record (the pre-computed hashCode prevents
that), this is one additional baby step in that direction.

joegallo committed:
Merge branch 'main' into cache-enrich-cache-key-hashcode

joegallo commented:
I'm closing this in favor of https://github.com/elastic/elasticsearch/pull/120256 (or something like it).
